lib/collection/src/collection/mod.rs

```diff
@@ -372,16 +372,30 @@ impl Collection {
         let current_state = replica_set.peer_state(&peer_id);
 
         // Validation:
-        // 0. Check that `from_state` matches current state
+        //
+        // 1. Check that peer exists in the cluster (peer might *not* exist, if it was removed from
+        //    the cluster right before `SetShardReplicaSet` was proposed)
+        let peer_exists = self
+            .channel_service
+            .id_to_address
+            .read()
+            .contains_key(&peer_id);
 
+        if !peer_exists {
+            return Err(CollectionError::bad_input(format!(
```

Comment:
We might still want to change replica state if replica already exists.
The only thing we don't want to do it to create new replica if none registered yet.

This might be only needed if something goes wrong, so I propose to include debug assert in this case

Reply:
> The only thing we don't want to do it to create new replica if none registered yet.

Not sure if this is true. We might actually rely on this somewhere. What we don't want to do, is create new replica *if peer does not exist*.

Reply:
Also, the fact that a lot of tests exploded already seems very sus... 😂

Reply:
> We might actually rely on this somewhere

I mean, we literally call `replica_set.ensure_replica_with_state(...)`, not just `replica_set.set_replica_state(...)`. Makes it reasonable to assume, that when `SetShardReplicaSet` is applied, it *creates* replica, if it does not exist yet.

Reply:
I'll change the check to `!peer_exists && !replica_exists` though. This way we only prevent (re)creating *new* replica, if peer is gone, but allow to change state of *existing* replica (even if peer is gone).

Reply:
I like that we still allow changing it if the replica already exists.

I wonder if we can clarify the semantics of this, and prevent calling this to spawn an unknown replica from thin air. It would require a multi-version effort though.